### ISSUE TITLE
Ephemeral Containers: Automatically taint pods with ephemeral containers

### ISSUE THREAD

**Issue by verb** [kubernetes/kubernetes#84353]
**What would you like to be added**: Automatically adding an annotation to pods in which an ephemeral container has been added.

**Why is this needed**: A common usage pattern for ephemeral containers is to perform some debugging action in a pod that changes its state in some fashion. Since the pod is no longer pristine, it should be marked as such so it could automatically be deleted.

**Comment by skilxn-go** [comment]
Can you assign this to me? I'm glad to contribute this, if it's still necessary.

**Comment by verb** [comment]
@skilxn-go Fantastic! I think it'd still be welcome and no one has started working on it yet. The first step will be getting agreement on the semantics. A great way to start would be to send a PR for [the ephemeral containers KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190212-ephemeral-containers.md) proposing how you'd like to solve this.

/assign @skilxn-go

**Comment by skilxn-go** [comment]
Thanks, I will give this proposal pr in two days.

**Comment by skilxn-go** [comment]
I think I shoud get aggrement with you first before I send PR. Can we taint pods with labels? I think maybe it's better to use labels for people can filter them out with Label Selector.

**Comment by verb** [comment]
@skilxn-go I agree, and it could be useful for automatically removing a tainted pod from a service, but I'm not aware of other instances of controllers automatically adding labels. I'm not sure how it would be received. It might be worth floating the idea at a SIG meeting.

**Comment by verb** [comment]
I've been thinking about this in the context of allowing deletion of ephemeral containers. Right now I'm thinking that the kubelet should track whether a pod has had an ephemeral container and indicate it in a `PodCondition`.

This would be default behavior for every cluster, and it would be simple to write a controller that applied labels based on the condition, allowing this part to be configurable per cluster.

**Comment by skilxn-go** [comment]
Yep, adding a `PodCondition` sounds good to me.
BTW, do we support adding ephemeral containers for static pods now?

**Comment by verb** [comment]
@skilxn-go not intentionally. Earlier I thought we could take advantage of this for testing, but thinking about it more we should probably prevent it if we can.

**Comment by MadhavJivrajani** [comment]
/assign
I can try taking a look at this again, I'm relatively new here which is why I'll probably take a few days getting the required context down if that's okay. Will keep you updated on this issue @verb 😄

**Comment by verb** [comment]
@MadhavJivrajani Great, thanks for taking up the feature! The KEP mentions this briefly and may be a good place to gather context:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/277-ephemeral-containers#identifying-pods-with-ephemeral-containers

**Comment by MadhavJivrajani** [comment]
Hey @verb! Quick update, I think I'm a little comfortable with the related topics and what needs to be done. I will start working on sending a PR to update the KEP. Will keep you posted, thanks! 😄 

Also, quick question: is `StartEphemeralContainer` as mentioned in the KEP, implemented yet? Because I was unable to find a reference to it, the only way I could find that emphemeral containers are started is: https://github.com/kubernetes/kubernetes/blob/bacab40a9f9380e968406e91483bdc81f1b41bb0/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L878

Is there anything I'm missing?

**Comment by verb** [comment]
@MadhavJivrajani Awesome, thanks! The reference to `StartEphemeralContainer` is likely out of date. iirc that was used in an earlier implementation. You're not missing anything. The link you posted is the only place ephemeral containers are started.

**Comment by verb** [comment]
> https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/277-ephemeral-containers#identifying-pods-with-ephemeral-containers

FYI @derekwaynecarr since it came up in the last sig-node meetup. This is how we're planning to allow for detection of non-pristine pods.

**Comment by MadhavJivrajani** [comment]
Hi @verb! Sorry that this issue has been so quite recently, I got around to thinking about what to change, and here's what I think would need to be done:

1. Add a new `PodCondition` here:
https://github.com/kubernetes/kubernetes/blob/51cbebab1fd2ec85842aba632b0efd0835254455/staging/src/k8s.io/api/core/v1/types.go#L2537

2. Invoke https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/api/v1/pod/util.go#L343
after creation of an ephemeral container:
https://github.com/kubernetes/kubernetes/blob/bacab40a9f9380e968406e91483bdc81f1b41bb0/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L878

Wdyt? I wanted to get your opinion on this before making a PR to update the KEP

**Comment by verb** [comment]
@MadhavJivrajani SGTM and thanks! Updating the KEP will allow us to discuss it with the kubelet experts. Since this will change the core API we'll need API review.

Some details to work out:

- In which direction does the condition work? That is, should it be ~IsTainted or ~IsPristine?
- How do we prevent the condition from being reverted once set? It should be a one-way gate.
- Should `kubectl exec` also set this condition? If so this might deserve a standalone KEP.
- Should this interact with pod readiness gates?

**Comment by MadhavJivrajani** [comment]
Hi @verb!

> In which direction does the condition work? That is, should it be ~IsTainted or ~IsPristine?

On creation of an ephemeral container (the first one on said Pod), adding a `PodCondition` called `IsTainted` is more intuitive imo

> How do we prevent the condition from being reverted once set? It should be a one-way gate.

If I understand correctly, if the following code is used to update/create `PodConditions` then if its called for the first time for an `IsTaintedDueToEphemeralContainer` type of `PodCondition` then, it will create it and for subsequent calls, since nothing in that `PodCondition` would change, it wouldn't be changed 

> Should this interact with pod readiness gates?

As far as I can tell, the `PodCondition` for ephemeral containers should not affect setting the Pod status to ready, if I understand correctly, this `PodCondition` would serve align more along the lines of providing a way for someone interacting with a cluster to know if a Pod has ever had an ephemeral container created to debug.

> Should kubectl exec also set this condition?

I don't think I fully understood this, could you please elaborate? 

---

I had a question regarding setting of the `PodCondition`: should this be set on *only* on successful creation of an ephemeral container or should be set regardless of whether its creation was successful or not?

**Comment by verb** [comment]
> > Should kubectl exec also set this condition?
> 
> I don't think I fully understood this, could you please elaborate?

Running a process in the container via `kubectl exec` "taints" the pod in the same way that creating an ephemeral container does. It's no longer pristine. Should `kubectl exec` also cause this condition to be set?

> 
> I had a question regarding setting of the `PodCondition`: should this be set on _only_ on successful creation of an ephemeral container or should be set regardless of whether its creation was successful or not?

I'm not sure, either. I think you'll need to decide which is better.

**Comment by MadhavJivrajani** [comment]
> Running a process in the container via kubectl exec "taints" the pod in the same way that creating an ephemeral container does. It's no longer pristine. Should kubectl exec also cause this condition to be set?

@verb thanks and apologies for inactivity, post [this](https://github.com/kubernetes/kubernetes/pull/103354#issuecomment-897451068), things became clearer. I think having a unified pod condition makes sense given the intention of ephemeral containers is to help with `kubectl exec` like tasks for distro-less images. However, I will try and bring this up in a sig-node meeting as well to try and get broader feedback and will update here.

> I'm not sure, either. I think you'll need to decide which is better.

I gave this some more thought and I think it should be set only on success and not on an attempt basis, mainly due to this example mentioned in the KEP:
> For example, cluster administrators may want to automatically apply a label or delete the pod after a configurable time. This may be accomplished by a controller watching for this PodCondition

**Comment by MadhavJivrajani** [comment]
FYI - https://github.com/kubernetes/enhancements/issues/3163
